src/core/server/saved_objects/migrations/types.ts

```diff
@@ -39,10 +39,10 @@ import { SavedObjectsMigrationLogger } from './core/migration_logger';
  *
  * @public
  */
-export type SavedObjectMigrationFn = (
-  doc: SavedObjectUnsanitizedDoc,
+export type SavedObjectMigrationFn<InputAttributes = unknown, MigratedAttributes = unknown> = (
+  doc: SavedObjectUnsanitizedDoc<InputAttributes>,
   context: SavedObjectMigrationContext
-) => SavedObjectUnsanitizedDoc;
+) => SavedObjectUnsanitizedDoc<MigratedAttributes>;
```

Comment:
So, my only interrogation around that is that atm migrations are all mutating and returning the input `doc` object, causing usage of these introduced generics difficult in existing migrations.

I.E

```ts
export const migrateToKibana660: SavedObjectMigrationFn<any, any> = doc => {
  if (!doc.attributes.hasOwnProperty('disabledFeatures')) {
    doc.attributes.disabledFeatures = [];
  }
  return doc;
};
```

Meaning that 

```ts
type InputAttrs = {
}

type OutputAttrs = {
    disabledFeatures: SomeType[];
}

export const migrateToKibana660: SavedObjectMigrationFn<InputAttrs, OutputAttrs> = doc => {
  if (!doc.attributes.hasOwnProperty('disabledFeatures')) {
    doc.attributes.disabledFeatures = [];
  }
  return doc;
};
```

Would fail to compile with something like `property disabledFeatures does not exists on type InputAttrs` on the `doc.attributes.disabledFeatures = [];` line.

Which is why I'm wondering if we should add the output properties on the input `doc`:

```ts
export type SavedObjectMigrationFn<InputAttributes = unknown, MigratedAttributes = unknown> = (
  doc: SavedObjectUnsanitizedDoc<InputAttributes & MigratedAttributes>,
  context: SavedObjectMigrationContext
) => SavedObjectUnsanitizedDoc<MigratedAttributes>;
```

Ideally I would have been using Partial, `doc: SavedObjectUnsanitizedDoc<InputAttributes & Partial<MigratedAttributes>>`, however the given example still fails with `disabledFeatures is optional on type SavedObjectUnsanitizedDoc<InputAttributes & Partial<MigratedAttributes>> but required on type SavedObjectUnsanitizedDoc<MigratedAttributes>` when returning the doc.

The other option being to decide that migration function should properly spread / construct the resulting migrated object without returning directly the input `doc`. @rudolf wdyt?

Reply:
> The other option being to decide that migration function should properly spread / construct the resulting migrated object without returning directly the input doc.

Given the high impact of even a small `cannot read property x of undefined` bug in a migration I'd say it validates the bit of extra effort and memory to construct a new output object :+1: 

We can leave existing migrations with `any` as they are, but then encourage new migrations to be written in this way. We should update our documentation to show this pattern and probably add this to the developer release notes to create some more visibility so that teams don't stumble on this.

Reply:
As a slightly more type-safe solution than any, teams can still use `InputAttributes & MigratedAttributes` as their `InputAttributes` generic paramater.

Reply:
SGTM, I'll keep the definition as it is right now then.